src/java/org/apache/cassandra/service/accord/AccordService.java

```diff
@@ -377,24 +371,33 @@ public synchronized void startup()
         node.commandStores().restoreShardStateUnsafe(topology -> configService.reportTopology(topology, true, true));
         configService.start();
 
-        long minEpoch = fetchMinEpoch();
-        if (minEpoch >= 0)
+        try
         {
-            for (long epoch = minEpoch; epoch <= metadata.epoch.getEpoch(); epoch++)
-                node.configService().fetchTopologyForEpoch(epoch);
+            // Fetch topologies up to current
+            List<Topology> topologies = fetchTopologies(0, metadata);
+            for (Topology topology : topologies)
+                configService.reportTopology(topology);
 
-            try
-            {
-                epochReady(metadata.epoch).get(DatabaseDescriptor.getTransactionTimeout(MILLISECONDS), MILLISECONDS);
-            }
-            catch (InterruptedException e)
-            {
-                throw new UncheckedInterruptedException(e);
-            }
-            catch (ExecutionException | TimeoutException e)
+            ClusterMetadataService.instance().log().addListener(configService.listener);
+            ClusterMetadata next = ClusterMetadata.current();
+
+            // if metadata was updated before we were able to add a listener, fetch remaining topologies
+            if (metadata != next)
```

Comment:
when i think about this this feels currently safe, but feels off... if for some reason in the future, if TCM ever recreates the `ClusterMetadata` object, then this logic breaks; should we just check the epochs don't match instead?

Reply:
This is a very common pattern, both with TCM and outside it, especially in Accord, so I assumed this is the way we commonly do it. 

But I don't mind comparing epochs either, switched to `if (!metadata.epoch.is(next.epoch))`.

Reply:
I can't speak for TCM, but in Accord this pattern has caused bugs... refactors break assumptions, and you have no clean way to find all the places where your refactor would break things... we had a bug a long time ago when we would do `if (cmd.txnId() == txnId)` and had to switch to `if (cmd.txnId().equals(txnId))`... a refactor broke the assumption that the objects won't be recreated, which then caused that condition to be `false` and do the wrong thing...